v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -196,20 +212,28 @@ public void processElement(ProcessContext c) {
       } catch (ChangeEventConvertorException ex) {
         outputWithTag(c, Constants.PERMANENT_ERROR_TAG, ex.getMessage(), spannerRec);
       } catch (SpannerException | IllegalStateException ex) {
-        if (!spannerRec.isRetryRecord()) {
-          retryableRecordCountMetric.inc();
+        outputWithTag(c, Constants.RETRYABLE_ERROR_TAG, ex.getMessage(), spannerRec);
+      } catch (com.mysql.cj.jdbc.exceptions.CommunicationsException
+          | java.sql.SQLTransientConnectionException ex) {
+        outputWithTag(c, Constants.RETRYABLE_ERROR_TAG, ex.getMessage(), spannerRec);
+      } catch (java.sql.SQLNonTransientConnectionException ex) {
+        if (ex.getMessage().contains("Server shutdown in progress")) {
+          outputWithTag(c, Constants.RETRYABLE_ERROR_TAG, ex.getMessage(), spannerRec);
+        } else {
+          outputWithTag(c, Constants.PERMANENT_ERROR_TAG, ex.getMessage(), spannerRec);
         }
+      } catch (ConnectionException ex) {
         outputWithTag(c, Constants.RETRYABLE_ERROR_TAG, ex.getMessage(), spannerRec);
       } catch (Exception ex) {
         LOG.error("Failed to write to source", ex);
         // we are only interested in the retryable errors
         // https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html
-        // Error 1452 and 1451
+        // Error 1452 and 1451 and transient communication errors
+
+        if (ex.getMessage().contains("a foreign key constraint fails")
+            || ex.getMessage().contains("No operations allowed after statement closed")
+            || ex.getMessage().contains("Got timeout reading communication packets")) {
```

Comment:
Why can't we use error codes here for matching?

Reply:
Thanks! Done